src/rpc_client/src/lib.rs

```diff
@@ -159,3 +157,43 @@ macro_rules! rpc_client_method_impl {
         )*
     }
 }
+
+// separate macro for connections against meta server to handle meta node failover
+#[macro_export]
+macro_rules! meta_rpc_client_method_impl {
+    ($( { $client:tt, $fn_name:ident, $req:ty, $resp:ty }),*) => {
+        $(
+            pub async fn $fn_name(&self, request: $req) -> $crate::Result<$resp> {
+                let req_clone = request.clone();
+                let response = self
+                    .$client
+                    .as_ref()
+                    .lock()
+                    .await
+                    .$fn_name(request.clone())
+                    .await;
+
+                // meta server is leader. Connection is correct
+                if response.is_ok() {
+                    return Ok(response.unwrap().into_inner());
+                }
+
+                // repeat request if we were connected against the wrong node
+                if self.do_failover_if_needed().await {
```

Comment:
If some RPC calls failed, it can already tell that if we should do failover based on the response code. And not all RPCs calls are retryable.